File: PE/PEFormat/DosHeader.ts
// <reference path="MZSignature.ts" />

module Mi.PE.PEFormat {
    export class DosHeader {
        mz: MZSignature;

        // Bytes on last page of file.
        cblp: number;

        // Pages in file.
        cp: number;

        // Relocations.
        crlc: number;

        // Size of header in paragraphs.
        cparhdr: number;

        // Minimum extra paragraphs needed.
        minalloc: number;

        // Maximum extra paragraphs needed.
        maxalloc: number;

        // Initial (relative) SS value.
        ss: number;

        // Initial SP value.
        sp: number;

        // Checksum.
        csum: number;

        // Initial IP value.
        ip: number;

        // Initial (relative) CS value.
        cs: number;

        // File address of relocation table.
        lfarlc: number;

        // Overlay number.
        ovno: number;

        // ulong
        res1: number;

        // OEM identifier (for e_oeminfo).
        oemid: number;

        // OEM information: number; e_oemid specific.
        oeminfo: number;

        reserved: number[]; // uint[5]

        // uint: File address of PE header.
        lfanew: number;

        toString() {
            var result = 
                "[" +
                (this.mz == MZSignature.MZ ? "MZ" :
                typeof this.mz == "number" ? (<number>this.mz).toString(16) + "h" :
                typeof this.mz) + "]" +

                ".lfanew=" +
                (typeof this.lfanew == "number" ? this.lfanew.toString(16) + "h" :
                typeof this.lfanew);

            return result;
        }
    }
}

File: Zoom.PE/Mi.PE/PEFormat/DosHeader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Mi.PE.PEFormat
{
    using Mi.PE.Internal;
    
    public sealed class DosHeader
    {
        public const int Size = 64;

        /// <summary> Bytes on last page of file. </summary>
        public ushort cblp;

        /// <summary> Pages in file. </summary>
        public ushort cp;

        /// <summary> Relocations. </summary>
        public ushort crlc;

        /// <summary> Size of header in paragraphs. </summary>
        public ushort cparhdr;

        /// <summary> Minimum extra paragraphs needed. </summary>
        public ushort minalloc;

        /// <summary> Maximum extra paragraphs needed. </summary>
        public ushort maxalloc;

        /// <summary> Initial (relative) SS value. </summary>
        public ushort ss;

        /// <summary> Initial SP value. </summary>
        public ushort sp;

        /// <summary> Checksum. </summary>
        public ushort csum;

        /// <summary>  Initial IP value. </summary>
        public ushort ip;

        /// <summary> Initial (relative) CS value. </summary>
        public ushort cs;

        /// <summary> File address of relocation table. </summary>
        public ushort lfarlc;

        /// <summary> Overlay number. </summary>
        public ushort ovno;

        public ulong res1;

        /// <summary> OEM identifier (for e_oeminfo). </summary>
        public ushort oemid;

        /// <summary> OEM information; e_oemid specific. </summary>
        public ushort oeminfo;

        public uint ReservedNumber0;
        public uint ReservedNumber1;
        public uint ReservedNumber2;
        public uint ReservedNumber3;
        public uint ReservedNumber4;

        /// <summary> File address of PE header. </summary>
        public uint lfanew;

        #region ToString
        public override string ToString()
        {
            return "[MZ].lfanew=" + this.lfanew.ToString("X") + "h";
        }
        #endregion
    }
}